lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -1600,27 +1600,77 @@ class ConvertAtenScaledDotProductAttentionOp
+    auto value = adaptor.getValue();
+    auto key = adaptor.getKey();
+
+    auto loc = op.getLoc();
+    auto collapseBatch = [&rewriter, loc](Value value) -> Value {
+      auto valueTy = cast<ShapedType>(value.getType());
+      if (valueTy.getRank() <= 3)
+        return value;
+
+      llvm::SmallVector<int64_t, 3> newShape(3, 1);
+      llvm::SmallVector<ReassociationIndices, 3> reassociation(3);
+      reassociation[1].push_back(valueTy.getRank() - 2);
+      reassociation[2].push_back(valueTy.getRank() - 1);
+      newShape[1] = valueTy.getDimSize(valueTy.getRank() - 2);
+      newShape[2] = valueTy.getDimSize(valueTy.getRank() - 1);
+
+      for (int i = 0, s = valueTy.getRank() - 2; i < s; ++i) {
+        reassociation.front().push_back(i);
+        if (valueTy.isDynamicDim(i) || newShape[0] == ShapedType::kDynamic) {
+          newShape[0] = ShapedType::kDynamic;
+          continue;
+        }
+        newShape[0] = newShape[0] * valueTy.getDimSize(i);
+      }
+
+      auto collapseTy = valueTy.clone(newShape);
+      return rewriter.create<tensor::CollapseShapeOp>(loc, collapseTy, value,
+                                                      reassociation);
+    };
+
+    query = collapseBatch(query);
+    key = collapseBatch(key);
+    value = collapseBatch(value);
+
     SmallVector<int64_t> outSizes(
-        adaptor.getQuery().getType().cast<ShapedType>().getShape());
+        query.getType().cast<ShapedType>().getShape());
     SmallVector<int64_t> valueSizes(
-        adaptor.getValue().getType().cast<ShapedType>().getShape());
+        value.getType().cast<ShapedType>().getShape());
     outSizes[outSizes.size() - 1] = valueSizes[valueSizes.size() - 1];
     SmallVector<Value> outSizesDynamic(
-        getTensorSizes(rewriter, op.getLoc(), adaptor.getQuery()));
-    outSizesDynamic[outSizesDynamic.size() - 1] = getTensorSizes(
-        rewriter, op.getLoc(), adaptor.getValue())[valueSizes.size() - 1];
+        getTensorSizes(rewriter, op.getLoc(), query));
+    outSizesDynamic[outSizesDynamic.size() - 1] =
+        getTensorSizes(rewriter, op.getLoc(), value)[valueSizes.size() - 1];
     Type outType = RankedTensorType::get(outSizes, elementType);
     Value output = createZeroInitTensor(rewriter, op.getLoc(), outSizesDynamic,
                                         elementType);
 
     // Overwrite with tm_tensor::attention
-    auto attention = rewriter.create<AttentionOp>(
-        op.getLoc(), outType,
-        SmallVector<Value>{adaptor.getQuery(), adaptor.getKey(),
-                           adaptor.getValue()},
-        SmallVector<Value>{output});
+    Value attention =
+        rewriter
+            .create<AttentionOp>(loc, outType,
+                                 SmallVector<Value>{query, key, value},
+                                 SmallVector<Value>{output})
+            .getResult()[0];
+
+    if (opTy != outType) {
+      llvm::SmallVector<ReassociationIndices, 3> reassociation(3);
```

Comment:
Can we re-use the reassociation from above? Just wondering when the expand reassociation will be different from the collapse one.

Reply:
Reworked to reuse the reassociation list for the collapses. Included some more dimension checks so operands must be the same rank.